src/fsharp/DependencyManager.Integration.fs

```diff
@@ -0,0 +1,182 @@
+
+open System
+open System.Reflection
+open System.IO
+open FSharp.Compiler.ErrorLogger
+open Internal.Utilities.FSharpEnvironment
+open FSharp.Compiler.DotNetFrameworkDependencies
+
+// Contract strings
+let dependencyManagerPattern = "*DependencyManager*.dll"
+let dependencyManagerAttributeName= "DependencyManagerAttribute"
+let resolveDependenciesMethodName = "ResolveDependencies"
+let namePropertyName = "Name"
+let keyPropertyName = "Key"
+
+module ReflectionHelper =
+    let assemblyHasAttribute (theAssembly: Assembly) attributeName =
+        try
+            CustomAttributeExtensions.GetCustomAttributes(theAssembly)
+            |> Seq.exists (fun a -> a.GetType().Name = attributeName)
+        with | _ -> false
+
+    let getAttributeNamed (theType: Type) attributeName =
+        try
+            theType.GetTypeInfo().GetCustomAttributes false
+            |> Seq.tryFind (fun a -> a.GetType().Name = attributeName)
+        with | _ -> None
+
+    let getInstanceProperty<'treturn> (theType: Type) propertyName =
+        try
+            let property = theType.GetProperty(propertyName, typeof<'treturn>)
+            if isNull property then
+                None
+            elif not (property.GetGetMethod().IsStatic)
+                 && property.GetIndexParameters() = Array.empty
+            then
+                Some property
+            else
+                None
+        with | _ -> None
+
+    let getInstanceMethod<'treturn> (theType: Type) (parameterTypes: Type array) methodName =
+        try
+            let theMethod = theType.GetMethod(methodName, parameterTypes)
+            if isNull theMethod then
+                None
+            else
+                Some theMethod
+        with | _ -> None
+
+    let stripTieWrapper (e:Exception) =
+        match e with
+        | :? TargetInvocationException as e->
+            e.InnerException
+        | _ -> e
+
+(* Shape of Dependency Manager contract, resolved using reflection *)
+type internal IDependencyManagerProvider =
+    abstract Name : string
+    abstract Key: string
+    abstract ResolveDependencies : scriptDir: string * mainScriptName: string * scriptName: string * packageManagerTextLines: string seq * tfm: string -> bool * string list * string list
+
+[<RequireQualifiedAccess>]
+type ReferenceType =
+| RegisteredDependencyManager of IDependencyManagerProvider
+| Library of string
+| UnknownType
+
+type ReflectionDependencyManagerProvider(theType: Type, nameProperty: PropertyInfo, keyProperty: PropertyInfo, resolveDeps: MethodInfo, outputDir: string option) =
+    let instance = Activator.CreateInstance(theType, [|outputDir :> obj|])
+    let nameProperty     = nameProperty.GetValue >> string
+    let keyProperty      = keyProperty.GetValue >> string
+
+    static member InstanceMaker (theType: System.Type, outputDir: string option) =
+        match ReflectionHelper.getAttributeNamed theType dependencyManagerAttributeName,
```

Comment:
In the original code, those were nested checks; it looks much better that way, but I believe the first check (attribute on the type) should be handled separately, there are potentially lots of types that will be coming in the assemblies, better to reject them on not having the attribute, then do the remaining checks.